beluga_amcl/include/beluga_amcl/filter_update_control/filter_update_control_mixin.hpp

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_AMCL__FILTER_UPDATE_CONTROL__FILTER_UPDATE_CONTROL_MIXIN_HPP_
+#define BELUGA_AMCL__FILTER_UPDATE_CONTROL__FILTER_UPDATE_CONTROL_MIXIN_HPP_
+
+#include <utility>
+#include <vector>
+#include <execution>
+
+#include <beluga/sensor.hpp>
+#include <beluga_amcl/tf2_sophus.hpp>
+
+#include <ciabatta/ciabatta.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a resampling policies poller.
+ */
+
+namespace beluga_amcl
+{
+
+/// @brief Public interface of the FilterUpdateControl filter customization mixin
+class FilterUpdateControlInterface
+{
+public:
+  using motion_update_type = Sophus::SE2d;
+  using sensor_update_type = std::vector<std::pair<double, double>>;
+
+  /// @brief Default destructor
+  virtual ~FilterUpdateControlInterface() = default;
+
+  /** @brief Execute a filter update step, conditioned by the installed resampling policies.
+   *  @param odom_to_base_transform Odom-to-base transform, for the motion update.
+   *  @param laser_scan Laser-scan information, for the sensor update.
+   *  @return true if the estimation result of the filter has been updated, false otherwise.
+   */
+  virtual bool update_filter(
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+
+  /** \overload
+   * It allows specifying a sequenced execution policy.
+   */
+  virtual bool update_filter(
+    std::execution::sequenced_policy exec_policy,
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+
+  /** \overload
+   * It allows specifying a parallel execution policy.
+   */
+  virtual bool update_filter(
+    std::execution::parallel_policy exec_policy,
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+};
+
+/// Resampling policy poller.
+/**
+ * \tparam Mixin The mixed-in type.
+ * \tparam UpdateFilterWhenMovingPolicy The resample on motion policy type.
+ * \tparam ResampleIntervalPolicy The resample interval policy type.
+ * \tparam SelectiveResamplingPolicy The selective resampling policy type.
+ */
+template<
+  typename Mixin,
```

Comment:
meta: Have you considered not making this class part of the particle filter mixin?

Or alternatively, and related to the following comment, have you considered moving just `UpdateFilterWhenMovingPolicy` out of the mixin?

> I still think the three policies (not the FilterUpdateControlMixin) can eventually go back into the main beluga library, but I don't think forcing a common abstraction on them will be meaningful.


To me it feels like a `ResampleIntervalPolicy` and `SelectiveResamplingPolicy` are core to the particle filter algorithm and should be provided by the main library. We can maybe think of a higher level interface that wraps the update sequence (update_motion, sample, update_sensor, reweight, evaluate resampling policy and resample) similar to what you did here providing the `update_filter` method.

When to call that method will depend on the Beluga AMCL code. That's where `UpdateFilterWhenMovingPolicy` comes into play.

Reply:
> meta: Have you considered not making this class part of the particle filter mixin?

Yes. However, the thing preventing this is that SelectiveResample currently needs access to the filter instance internals through a reference-to-this.

> To me it feels like a ResampleIntervalPolicy and SelectiveResamplingPolicy are core to the particle filter algorithm and should be provided by the main library.

Agreed, I just don't yet know what's the best way to integrate them; all three samples that we have are different in their own way. I think we still need to find how to integrate "shards" (reusable modules smaller than a mixin) to the library if we don't want to just make them small mixins that just add one or two functions (I'm not against this, though; the only limitations I can think of this would impose are a) further namespace inflation within an assembled mixin, b) impossibility to instantiate more than once if needed).

I'd rather defer solving this until after we have solved the performance and behavioral compatibility issue.

> We can maybe think of a higher level interface that wraps the update sequence (update_motion, sample, update_sensor, reweight, evaluate resampling policy and resample) similar to what you did here providing the update_filter method.

I think it makes sense. I would add this along with at least some examples where we actually use in a full filter to create other filters ("beluga howto"), because as of now the only instantiations of the filter that we have (the integration tests and `beluga_amcl`) will all use the nav2-like update control, so any other alternative we provide may go unnoticed and unused. It's going to become more important when `beluga` has other applications besides `beluga_amcl`.

It also makes sense as a way to narrow the external interface to the assembled mixin. Currently the user has access to any method that's present in any of the declared interfaces, which is probably far more than they need.

Reply:
> Yes. However, the thing preventing this is that SelectiveResample currently needs access to the filter instance internals through a reference-to-this.

Ah! Makes sense.

> I'd rather defer solving this until after we have solved the performance and behavioral compatibility issue.

Okay. Sounds good to me.

> I would add this along with at least some examples where we actually use in a full filter to create other filters ("beluga howto"), because as of now the only instantiations of the filter that we have (the integration tests and beluga_amcl) will all use the nav2-like update control, so any other alternative we provide may go unnoticed and unused.

Good point!